### ISSUE TITLE
batch correction

### ISSUE THREAD

**Issue by JBreunig** [KrishnaswamyLab/MELD#50]
Brilliant work on the paper and package! I also enjoyed the linked Zoom presentation.

My typical workflow is to concat loom files and do Scanpy QC followed by Harmony integration.  

I see how you go from an anndata object in this tutorial:  https://github.com/KrishnaswamyLab/MELD/blob/master/notebooks/MELD_Quickstart.ipynb

Do you have any recommendations for a batch correction workflow for going from an anndata object and running MELD?

**Comment by dburkhardt** [comment]
Glad to hear it was helpful! Can you please clarify what part you need help with?

If you already have batch-corrected data, then you can simply pass that matrix to MELD. It's important to ensure before running MELD, that the batch normalization was applied properly. Make sure you can still see differences between treatment and control conditions in a PHATE embedding, for example. Or you could check to see that there is some differential expression for a couple of markers you expect to be different across treatment and control conditions.

There's also some advice in Supplementary Note 3, which is unfortunately buried in the supplement. I'll copy it here:

**Supplementary Note 3: Applying MELD analysis to single cell datasets with a batch effect**

When jointly analyzing single cell datasets collected in different samples, difficulty may arise due to systematic changes in gene expression profiles between biologically equivalent cells [7]. These changes may be technical in nature (e.g. differences in the reverse transcription efficiency during library preparation) or biological (e.g. changes in sample preparation cause unexpected changes in biological state of otherwise equivalent cells). Regardless of the cause, the unifying feature of batch effects is that they confound the analysis a given research wants to perform. As such, it is unsurprising that dozens of batch normalization tools have been developed for single cell data [12]. However, it is important to emphasize that what constitutes a batch effect is dependent on the biological question in which a researcher is interested. Some analysts might be uninterested in variation caused by a change in cell media composition between samples, but other researchers might want to study these differences. Batch normalization tools have no way to know what variation is biologically relevant to the specific hypotheses of a given experiment and thus risk removing meaningful experimental signal when ”correcting” measured values. This is problematic for analysis using MELD, because the goal of the toolkit is to quantify the differences that exist between samples without regard for the specific interests of given hypothesis. As such, we do not recommend using batch correction along the experimental axis (i.e. between treatment and control conditions) before running MELD. However, recognizing that in some cases batch correction is essential, we describe several considerations for performing MELD analysis on batch-corrected data.   
  
For the MELD algorithm to accurately estimate relative likelihood for each sample, we assume that the graph learned from single cell data approximates the underlying cell state manifold. In the Methods we describe the use of an anisotropic kernel that normalizes for varying sampling density across cell states. However, some batch correction methods, such as mutual nearest neighbors [13], rely on the construction of a graph with artificially inflated weights between nodes from different samples. This graph no longer models the cell states an experiment measured, but rather enforces similarities between cells based on the heuristic of the chosen normalization model. We provide no theoretical guarantees that a graph learned from batch corrected data will accurately model the underlying probability densities of each condition.
  
In practice when analyzing islet cells collected from multiple donors, that applying batch correction methods across the donor label improves our ability to capture a signal of IFNg stimulation. It is important to note that in this case, batch correction applied to a label that is orthogonal to the experimental axis. We have no examined the accuracy of the MELD algorithm when batch correction is applied between experimental and control samples, although it is our expectation that this will likely remove biological signal. We recommend any user considering applying batch correction methods prior to running MELD analysis follow these steps:  

1. To determine if a batch effect exists, confirm that cells from one sample are not finding appropriate neighbors in another following the strategy outlined by Buttner et al. [7].  
2. To characterize the effect, identify which genes change the most between the samples  
3. Confirm that the genes that are different are not relevant to the biological question under investigation  
4. Apply batch correction  
5. Confirm that relevant biological differences are still present using MELD analysis  
6. If the biological differences are not present, repeat from step 1 with less batch correction. If you hit your personal recursion limit, consider that you don’t actually want to do batch correction
7. If biological differences are present, then confirm that previous batch effect has been corrected and proceed to downstream analysis

**Comment by JBreunig** [comment]
Thanks, I did miss that note!

I guess my specific question is whether there is any way to incorporate Harmony in terms of batch correction or if I am limited to batch correction that outputs a "corrected" matrix (e.g. limma, ComBat, Seurat, Scanorama)?

**Comment by dburkhardt** [comment]
I haven't used Harmony before, but scanning their paper, it looks like the output is an embedding of cells. You could use this embedding as input to the graph for MELD just as you might for PHATE or UMAP. I haven't tried this before though. I'd be interested to see your results!

**Comment by JBreunig** [comment]
Ok, I'll give it a shot, thanks!

**Comment by Rohit-Satyam** [comment]
@JBreunig were you able to integrate MELD with Harmony?

**Comment by JBreunig** [comment]
It has been a few months but I basically just fed the Harmony corrected object with calculated UMAP into MELD.  MELD had no problems dealing with it.

**Comment by Rohit-Satyam** [comment]
Could you please share a exampler code snippet.

**Comment by JBreunig** [comment]
I used the standard scanpy workflow and then used the harmony batch correction.  Then, I subsetted to a few particular clusters and ran the following with that subset at 'adata' in memory:

import os
import pandas as pd
import numpy as np
import scprep
import meld
import scprep
import phate
import meld
np.random.seed(42)
from matplotlib import pyplot as plt


%load_ext autoreload
%autoreload 2
meld_op = meld.MELD()
sample_densities = meld_op.fit_transform(adata, sample_labels=adata.obs['sample_labels'])

fig, axes = plt.subplots(2,2, figsize=(13,7))

for i, sample in enumerate(np.unique(sample_densities.columns)):
    ax = axes.flatten()[i]
    scprep.plot.scatter2d(adata.obsm['X_umap'], title=sample,
                          c=sample_densities[sample], cmap='inferno', 
                          vmin=0, vmax=0.00015, ax=ax, ticks=False)
    
fig.tight_layout()

Xgene_likelihoods = []
for replicate in 'AB':
    # Calculate the ratio of sample densities across matched treatment and control samples 
    # (e.g. chdA vs tyrA)
    Xgene_likelihoods.append(
        meld.normalize_densities(
            sample_densities[['het{}'.format(replicate),'Xgeneko{}'.format(replicate)]]
        )
    )

# Only need to keep the likelihood of the chd conditions
Xgeneko_likelihoods = pd.concat(Xgene_likelihoods, axis=1)[['XgenekoA', 'XgenekoB']]

fig, axes = plt.subplots(1,2, figsize=(12,4))

for i, sample in enumerate(Xgeneko_likelihoods.columns):
    ax = axes.flatten()[i]
    scprep.plot.scatter2d(adata.obsm['X_umap'], title=sample,
                          c=Xgeneko_likelihoods[sample], cmap=meld.get_meld_cmap(),
                          vmin=0, vmax=1,
                          ax=ax, ticks=False)
    
fig.tight_layout()


fig, axes = plt.subplots(1,3, figsize=(12,4))

scprep.plot.scatter2d(adata.obsm['X_umap'], title='Mean\nRelative Likelihood',
                      c=Xgeneko_likelihoods.mean(axis=1), cmap=meld.get_meld_cmap(),
                      vmin=0, vmax=1, ax=axes[0], ticks=False)

scprep.plot.scatter2d(adata.obsm['X_umap'], title='Std.\nRelative Likelihood',
                      c=Xgeneko_likelihoods.std(axis=1), cmap='inferno',
                      vmin=0, ax=axes[1], ticks=False)

scprep.plot.histogram(Xgeneko_likelihoods.std(axis=1), title='Std.\nRelative Likelihood',
                      ax=axes[2])
axes[2].axvline(0.08, color='r', linewidth=0.5)

fig.tight_layout()

replicate_likelihoods = [meld.normalize_densities(sample_densities[['hetA', 'hetB' ]]),
                         meld.normalize_densities(sample_densities[['XgenekoA', 'XgenekoB' ]])]
replicate_likelihoods = pd.concat(replicate_likelihoods, axis=1)

fig, axes = plt.subplots(1,2, figsize=(8,4))

for i, sample in enumerate('AB'):
    ax = axes.flatten()[i]
    if i < 3:
        batch = 'Xgeneko'
    else:
        batch = 'het'
    if i % 3 == 0:
        ylabel = batch
    else:
        ylabel = None
    scprep.plot.scatter2d(adata.obsm['X_umap'], title=sample, ylabel=ylabel,
                          c=replicate_likelihoods['{}{}'.format(batch, sample)], cmap='inferno',
                          vmin=0, vmax=1,
                          ax=ax, ticks=False)

fig.tight_layout()


sce.tl.phate(adata, k=5, a=20, t=150)
sce.pl.phate(adata)


data_pca = scprep.reduce.pca(data_sqrt)

phate_op = phate.PHATE(knn=10, decay=10, n_jobs=-1)
data_phate = phate_op.fit_transform(data_pca)
